File: Linked_list/linked_list.py


class Node:
    def __init__(self,data=None):
        self.data=data
        self.next=None

class linkedList:
    def __init__(self):
        self.head=Node()
    
    def insert(self,data):
        self.new_node=Node(data)
        if self.head.next == None:
            self.head.next=self.new_node
        else :
            node=self.head.next
            while node.next !=None:
                node=node.next
            node.next=self.new_node

    def display(self):
        node=self.head.next
        while node.next !=None:
            print(node.data)
            node=node.next
        print(node.data)

    def delete(self,data):
        node =self.head.next
        #delete first node 
        if node.data ==data:
            next=node.next
            self.head.next=next
            del node
        else:
            prev_node=self.head
            while node.data !=data:
                node = node.next
                prev_node=prev_node.next
            prev_node.next=node.next
            del node
        
    def display_using_Rec(self,head):
        if (head.next==None):
            return 
        print(head.next.data)
        self.display_using_Rec(head.next)
        
  def recursive_reverse(self, temp):
    # recuse untill we reach the last node and make the head point to it
    # and then for each node the node after it will point to it ,and the next
    # of the node it self will be None.
    if temp.next == None:
        self.head.next = temp
        return
    self.reverse(temp.next)
    temp.next.next = temp
    temp.next = None
    

File: Linked_list/linked_list.cs
class Node<T>
        {
        public T data;
        public Node<T> next;
        public Node<T> head = null;

        public void Insert(T data)
        {
            Node<T> temp = new Node<T>();
            temp.data = data;
            temp.next = null;
            if (head==null)
            {
                head = temp;
                return;

            }
            Node<T> temp2 = head;
            while(temp2.next!=null)
            {
                temp2 = temp2.next;
            }
            temp2.next = temp;
        }
        void recursive_display(ref Node<T> head)
        {
            if (head == null)
            {
                return;
            }
            Console.WriteLine(head.data);
            recursive_display(ref head.next);
        }
        public void Display()
        {

            recursive_display(ref head);
        }
        public void Delete(T data)
        {
            if (head.data.Equals(data))
            {
                Node<T> temp2 = head.next;
                head= temp2;
                return;
            }
            Node<T> next_node = head;
            while (next_node.next != null && !EqualityComparer<T>.Default.Equals(next_node.next.data ,data) )
            {
                next_node = next_node.next;
            }
            if (next_node.next !=null)
                next_node.next = next_node.next.next;
        }

        public bool Find(T data)
        {
            Node<T> temp = head;
            while (temp.next !=null && !EqualityComparer<T>.Default.Equals(temp.data,data))
            {
                temp = temp.next;
            }
            if (EqualityComparer<T>.Default.Equals(temp.data ,data))
                return true;
            else
                return false;


        }
    }
